README.md
# Pie42's FRHD Mods
This is a place to store and compile all of the mods I have made for FRHD.

The purpose of this repository is to have all of my mods in one place, let me put out changes (and have them be propogated automatically), and make it easier to find all of my mods.

They can all be insalled by clicking on the link at the header. If you have tampermonkey instaled, it will open a new window were you can click "Install"
![image](https://user-images.githubusercontent.com/29359616/120890439-a5c5b880-c5d0-11eb-931b-e1fc7542a7da.png)


# Mods
## <a href="https://github.com/Pie42/frhd-mods/raw/main/Track%20Stat%20Mod.user.js">Star Finder</a>
Creates arrows pointing to all of the uncollected stars in a track.

![image](https://user-images.githubusercontent.com/29359616/120866664-6e1e2880-c55e-11eb-89d2-1bd2ba14432a.png)
<details>
  <summary>Changelog</summary>
  
  - Created mod. (<a href="https://github.com/Pie42/frhd-mods/commit/180127413b41b08df6022d113187fd99ed5dc7b0">1801274</a>)
</details>

---
## <a href="https://github.com/Pie42/frhd-mods/raw/main/FRHD%20Input%20Display.user.js">Input Display 1</a>

The input mods add display to show you what buttons you or the active/highlighted ghost are pressing. This can be good for recording gameplay and watching mindblowing ghosts.

![image](https://user-images.githubusercontent.com/29359616/120868733-acb5e200-c562-11eb-833d-e899055f0c35.png)

<details>
  <summary>Changelog</summary>

  - Created mod. (<a href="https://github.com/Pie42/frhd-mods/commit/180127413b41b08df6022d113187fd99ed5dc7b0">1801274</a>)
  - Added pressed input color customization. (Thanks <a href="https://www.twitter.com/revelcw">@revelcw</a>!) (<a href="https://github.com/Pie42/frhd-mods/commit/4fe12f5a99c5fb12168e2e300cfcc07691561d4b">4fe12f</a>)
</details>

---

## <a href="https://github.com/Pie42/frhd-mods/raw/main/Track%20Stat%20Mod.user.js">Track Stat</a>
Shows you some info on the current track for a few seconds at the beginning of the track.

![image](https://user-images.githubusercontent.com/29359616/120873445-1ee0f380-c570-11eb-8434-f238a4e96425.png)
<details>
  <summary>Changelog</summary>

  - Created mod. (<a href="https://github.com/Pie42/frhd-mods/commit/180127413b41b08df6022d113187fd99ed5dc7b0">1801274</a>)
</details>

---

## <a href="https://github.com/Pie42/frhd-mods/raw/main/Better%20Races.user.js">Better Races</a>
This mod adds a better race info to the top left of the screen. This includes adding an amount of stars for all ghosts.

![image](https://user-images.githubusercontent.com/29359616/120867702-76776300-c560-11eb-9a2d-8069f86be479.png)
<details>
  <summary>Changelog</summary>

  - Created mod. (<a href="https://github.com/Pie42/frhd-mods/commit/180127413b41b08df6022d113187fd99ed5dc7b0">1801274</a>)
</details>


CHANGELOG.md
## 06/04/2021
Added pressed input color customization. (Thanks @revelcw!)
([4fe12f](https://github.com/Pie42/frhd-mods/commit/4fe12f5a99c5fb12168e2e300cfcc07691561d4b))


Better Races.user.js
// ==UserScript==
// @name         Better Races
// @namespace    http://tampermonkey.net/
// @version      0.2
// @description  try to take over the world!
// @author       Pie42
// @match        https://www.freeriderhd.com/*
// @grant        none
// ==/UserScript==

//unfortunately, this script has much less customization, but you could still change it by changing the numbers for v.x, v.y, f.x, and f.y, or by changing
//'GameSettings.raceColors' to have different colors.

function s(t) { //this is a direct copy of the time parsing function in the game
            t = parseInt(t, 10);
            var s = Math.floor
              , e = s(t / 6e4)
              , i = (t - 6e4 * e) / 1e3;
            return i = i.toFixed(2),
            10 > e && (e = e),
            10 > i && (i = "0" + i),
            e + ":" + i
        }
function raceUpdate() { //this is the function that makes races fancy and cool
    var h = GameManager.game.currentScene.settings.drawFPS; //needed for parsing
    for (let i in window.uselessRaceThing) { //updating every race
        var cGhost = window.uselessRaceThing[i][0].text, //getting the current race to update,
            split = cGhost.split(' '),                   //that race's text split by spaces,
            dname = (split.length == 2) ? split[0] : split[0].slice(0,-1),                //the username in that race(the slice is necessary because the username is stored as {name}:),
            time = window.uselessRaceThing[i][1],        //the time of that race, also stored in that array,
            stars, rGhost;                               //and just setting up some variables for later use
        for (let j in GameManager.game.currentScene.playerManager._players) { //going through the players
            let cPlayer = GameManager.game.currentScene.playerManager._players[j];
            if (cPlayer._ghost && cPlayer._user.d_name == dname) { //if the player is a ghost and the name is the name in the display,
                stars = cPlayer._powerupsConsumed.targets.length;  //get how many stars that ghost has collected
                rGhost = cPlayer;                                  //get that entire player for some reason
                break;
            } //if the ghost hasn't finished, the display is set to an updated version of the form the races start out in
        } //if the ghost has finished, the display is set to {name} + a zero width space + ' finished'
        //the zero width space was the easiest way i could find to fix a bug that was occuring with names being deleted 1 character at a time
        var newT = (stars == GameManager.game.currentScene.track.targetCount) ? dname + ' finished!' : dname + ': ' + stars.toString() + ' / ' + GameManager.game.currentScene.track.targetCount.toString() + ' - ' + s((parseInt(time) - GameManager.game.currentScene.ticks) / h * 1e3);
        window.uselessRaceThing[i][0].text = newT;
    }
}
window.uselessRaceThing = []; //very useful race thing
function replace() {
    GameManager.game.currentScene.raceTimes.addRace = function(t, e) {
            if (this.raceCount < this.maxRaces) {
                var i = this.scene
                  , n = i.game
                  , r = (n.pixelRatio,
                t.user)
                  , o = t.race
                  , a = i.settings
                  , h = a.drawFPS
                  , l = r.color
                  , c = "helsinki"
                  , u = new createjs.Container
                  , v = new createjs.Text('>>', "30px Roboto Medium", l) //only in roboto because helsinki doesn't have the '>' character
                  , f = new createjs.Text(r.d_name + ': 0 / ' + i.track.targetCount.toString() + ' - ' + s(parseInt(o.run_ticks) / h * 1e3), "30px " + c, l); //setting up the display
                v.alpha = 0; //you can just remove the pointer initially, it will be highlighted if needed
                v.x = 1, //uh these numbers work, but feel free to mess around with them
                v.y = 9; //with these numbers you could put the display on the right side, but i couldn't find anywhere that looked good to me,
                f.x = 50,//so that's why it's on the left still
                f.y = 9;
                window.uselessRaceThing.push([f, o.run_ticks]); //very important line, since it lets it get updated
                u.addChild(v, f), //this is just stuff from the default script
                a.mobile ? u.x = e * this.mobileRaceXOffset : (u.x = -2,
                u.y = e * this.raceYOffset),
                this.raceList.push(u),
                this.container.addChild(u),
                this.raceCount++
            }
        }
    GameManager.game.currentScene.raceTimes.highlightRace = function(t) {
            if (GameManager.game.currentScene.raceTimes.highlightedRace !== GameManager.game.currentScene.raceTimes.raceList[t]) {
                GameManager.game.currentScene.raceTimes.unhighlightRace();
                var e = GameManager.game.currentScene.raceTimes.raceList[t].children[0]; //this is the only change i made to this
                e.alpha = 1, //basically, this sets the alpha of the '>>' string to 1, so it's visible
                GameManager.game.currentScene.raceTimes.highlightedRace = e
            }
        }
    GameManager.game.currentScene.raceTimes.unhighlightRace = function() {
            GameManager.game.currentScene.raceTimes.highlightedRace && (GameManager.game.currentScene.raceTimes.highlightedRace.alpha = 0, //hiding it again
            GameManager.game.currentScene.raceTimes.highlightedRace = null)
        }
    for (var race in GameManager.game.currentScene.races) { //now, because this script isn't inserted, once all of the functions are changed,
        GameManager.game.currentScene.addRaces([GameManager.game.currentScene.races[race]]); //it needs to re-add all of the races.
    }
}
var track = 0; //making sure it works still when you go to a different track
window.setInterval(function () {
    if ($("#track-data").data("t_id") != track) {
        track = $("#track-data").data("t_id");
        window.uselessRaceThing = [];
        function rInterval() {
            window.clearInterval(v)
        }
        var v = window.setInterval(function() {
            if (GameManager != undefined && GameManager.game != undefined) {
                rInterval();
                replace();
                createjs.Ticker.addEventListener("tick", raceUpdate);
            }
        }, 250)
    }
}, 500)

Track Stat Mod.user.js
// ==UserScript==
// @name         Track Stat Mod
// @namespace    http://tampermonkey.net/
// @version      0.1
// @description  try to take over the world!
// @author       Pie42
// @include      *freeriderhd.com/t/*
// @grant        none
// ==/UserScript==

//time (in frames) for the stats to fade away
var fadeTime = 250,
    //color of the stats
    color = [0, 0, 0],
    //font (will probably default to monospace if you enter something invalid)
    font = 'Monospace';
function s(t) { //this is a direct copy of the time parsing function in the game
    t = parseInt(t, 10);
    var s = Math.floor
    , e = s(t / 6e4)
    , i = (t - 6e4 * e) / 1e3;
    return i = i.toFixed(2),
        10 > e && (e = e),
        10 > i && (i = "0" + i),
        e + ":" + i
}

function inject() {
    var a = Array.from(document.getElementsByClassName('stat-container')).slice(0,-1);
    a = a.map(i=>i.children[0].innerHTML.trim());
    var f = Array.from(document.getElementsByClassName('track-leaderboard-race-row')),
        ff = f.filter(i=>i.dataset.type=='global'&&i.dataset.run_time!='false'),
        g = ff.map(i=>{let k = i.dataset.run_time,l=k.split(':'),m=l[1].split('.'),n=0;n+=(parseInt(m[1])*.3)|0;n+=parseInt(m[0])*30;n+=parseInt(l[0])*1800;return ((n+1)||Infinity)-1}),
        h = g.reduce((i,j)=>i+j,0)/g.length,
        q = Math.sqrt(g.reduce((i,j,k)=>((i*k)+((h-j)**2))/(k+1),0)),
        z = s((1000/GameManager.game.currentScene.settings.drawFPS)*q|0),
        y = [[`Players Completed: ${a[0]}`, (((+a[0]/10)+1)||2)-1],[`Completion Rate:   ${a[2]}`, +a[2]||0],[`Average Time:      ${a[1]}`,0.5],[`LB Average:        ${s(h*1000/GameManager.game.currentScene.settings.drawFPS)}`,Math.min(2/(h**0.2), 1)],[`LB Std Dev:        ${z}`,Math.min(4/(q**0.5), 1)]],
        x = document.createElement('canvas'),
        w = x.getContext('2d');
    w.textBaseline = 'top';
    window.tsmodel = createjs.Ticker.addEventListener('tick', function() {if (GameManager.game.currentScene.ticks < fadeTime) {let fs = Math.sqrt(Math.sqrt(GameManager.game.canvas.width**2 + GameManager.game.canvas.height**2)) / 2; x.width = Math.max(...y.map(i=>w.measureText(i[0]).width)); x.height = fs * y.length; w.textBaseline = 'top'; w.font = `${fs}px ${font}`; w.clearRect(0, 0, x.width, x.height); w.fillStyle = `rgba(250,250,250,${Math.max(1-GameManager.game.currentScene.ticks / fadeTime, 0)/2}`; w.fillRect(0,0,x.width,x.height); /*w.fillStyle = `rgba(${color[0]}, ${color[1]}, ${color[2]}, ${Math.max(1-GameManager.game.currentScene.ticks / fadeTime, 0)}`;*/ w.globalAlpha = Math.max(1-GameManager.game.currentScene.ticks / fadeTime, 0); for (var i = 0; i < y.length; (w.fillStyle = `rgb(${(1-y[i][1])*255},${y[i][1]*255},0)`),w.fillText(y[i][0], 0, i++ * fs)); GameManager.game.canvas.getContext('2d').drawImage(x, (GameManager.game.canvas.width - x.width)/2, GameManager.game.canvas.height - (fs * y.length));}});
    //console.log(a,f,ff,h,z,y,x,w);
}

var track = 0;
window.setInterval(function () {
    if ($("#track-data").data("t_id") != track || window.tsmodel == undefined) {
        track = $("#track-data").data("t_id");
        function rInterval() {
            window.clearInterval(v)
        }
        var v = window.setInterval(function() {
            if (GameManager != undefined && GameManager.game != undefined) {
                window.tsmodel!=undefined && createjs.Ticker.removeEventListener('tick', window.tsmodel);
                inject();
                rInterval();
            }
        }, 250)
    }
}, 500)

FRHD Input Display.user.js
// ==UserScript==
// @name         FRHD Input Display
// @namespace    http://freeriderhd.com/
// @version      0.91
// @description  try to take over the world!
// @author       Pie42
// @include      https://www.freeriderhd.com*
// @grant        none
// ==/UserScript==

/*
  Short explanation: The way this works is that the main drawing function first hooks itself up to the game's main update function,
  because that's the best way to make it draw correctly. After that, every time the game ticks, the input display is drawn onto the game canvas.
  To get button highlighting, each drawing function checks the current player's '_gamepad.downButtons' to see if the current button is being held down
  (whether it's by the current player or the current replay). If the current button is being held down, the box is filled in.

  This input display will automatically highlight the current player, meaning that if you're watching a ghost, the display will show the ghost's
  inputs, and if you're playing, it will show your inputs. I decided not to include 'enter' and 'backspace' because it's pretty easy to tell if those
  are being held. Enjoy!

  Note that you can customize this to your heart's content. I have indicated the main customization things, but feel free to change whatever(just don't
  blame me if it breaks).
*/

function drawDisplay() {
    var container = document.getElementById('game-container'),
        children = container == void 0 ? undefined : container.children,
        canvas = children == void 0 ? undefined : children[1];
    if (canvas != void 0) {
        var ctx = canvas.getContext('2d');
        var cPlayer =
            GameManager.game.currentScene.playerManager._players[
                GameManager.game.currentScene.camera.focusIndex
            ],
            cGamepad = cPlayer._gamepad,
            h = canvas.height,
            w = canvas.width;

        //main customization variables
        //to customize a symbol, find the comment stating that key(keys are listed left to right and top to bottom), and then change the drawing done there.
        var bsize = h / 20, //size of the boxes
            dist = h / 40, //distance from the edge of the screen
            lcolor = '#000000', //default line color
            ldcolor = '#fff', //line color of currently pressed inputs
            fcolor = '#000', //fill color of currently pressed inputs
            lwidth = 2; //width of the lines

        //sy and sx can also be customized. right now, they are set to put the input display in the lower left corner, but to move it to the top, you can use:
        //var sy = dist,
        //instead of the line that is currently there.
        var sy = h - dist - 2 * bsize,
            //to move the display to the right, you can use
            //sx = (w - dist) - (3 * bsize),
            //instead of the line that is currently there.
            sx = dist,
            tempx,
            tempy;
        ctx.strokeStyle = lcolor;
        ctx.fillStyle = fcolor;
        ctx.lineWidth = lwidth;

        //up arrow key
        ctx.strokeStyle = lcolor;
        tempx = sx + bsize;
        tempy = sy;
        ctx.beginPath();
        ctx.lineTo(tempx, tempy);
        ctx.lineTo(tempx, tempy + bsize);
        ctx.lineTo(tempx + bsize, tempy + bsize);
        ctx.lineTo(tempx + bsize, tempy);
        ctx.lineTo(tempx, tempy);
        cGamepad.downButtons.up && !cPlayer._crashed && ctx.fill();
        ctx.stroke();

        //change this to change the way the symbol is drawn
        if (cGamepad.downButtons.up && !cPlayer._crashed) {
            ctx.strokeStyle = ldcolor;
        }
        ctx.beginPath();
        ctx.lineTo(tempx + (1 * bsize) / 6, tempy + (2 * bsize) / 3);
        ctx.lineTo(tempx + (1 * bsize) / 2, tempy + (1 * bsize) / 3);
        ctx.lineTo(tempx + (5 * bsize) / 6, tempy + (2 * bsize) / 3);
        ctx.stroke();

        //z key
        ctx.strokeStyle = lcolor;
        tempx = sx + 2 * bsize;
        tempy = sy;
        ctx.beginPath();
        ctx.lineTo(tempx, tempy);
        ctx.lineTo(tempx, tempy + bsize);
        ctx.lineTo(tempx + bsize, tempy + bsize);
        ctx.lineTo(tempx + bsize, tempy);
        ctx.lineTo(tempx, tempy);
        cGamepad.downButtons.z && !cPlayer._crashed && ctx.fill();
        ctx.stroke();

        //change this to change the way the symbol is drawn
        if (cGamepad.downButtons.z && !cPlayer._crashed) {
            ctx.strokeStyle = ldcolor;
        }
        ctx.beginPath();
        ctx.lineTo(tempx + bsize / 6, tempy + bsize / 6);
        ctx.lineTo(tempx + (5 * bsize) / 6, tempy + bsize / 6);
        ctx.lineTo(tempx + bsize / 6, tempy + (5 * bsize) / 6);
        ctx.lineTo(tempx + (5 * bsize) / 6, tempy + (5 * bsize) / 6);
        ctx.stroke();

        //left key
        ctx.strokeStyle = lcolor;
        tempx = sx;
        tempy = sy + bsize;
        ctx.beginPath();
        ctx.lineTo(tempx, tempy);
        ctx.lineTo(tempx, tempy + bsize);
        ctx.lineTo(tempx + bsize, tempy + bsize);
        ctx.lineTo(tempx + bsize, tempy);
        ctx.lineTo(tempx, tempy);
        cGamepad.downButtons.left && !cPlayer._crashed && ctx.fill();
        ctx.stroke();

        //change this to change the way the symbol is drawn
        if (cGamepad.downButtons.left && !cPlayer._crashed) {
            ctx.strokeStyle = ldcolor;
        }
        ctx.beginPath();
        ctx.lineTo(tempx + (2 * bsize) / 3, tempy + (1 * bsize) / 6);
        ctx.lineTo(tempx + (1 * bsize) / 3, tempy + (1 * bsize) / 2);
        ctx.lineTo(tempx + (2 * bsize) / 3, tempy + (5 * bsize) / 6);
        ctx.stroke();

        //down key
        ctx.strokeStyle = lcolor;
        tempx = sx + bsize;
        tempy = sy + bsize;
        ctx.beginPath();
        ctx.lineTo(tempx, tempy);
        ctx.lineTo(tempx, tempy + bsize);
        ctx.lineTo(tempx + bsize, tempy + bsize);
        ctx.lineTo(tempx + bsize, tempy);
        ctx.lineTo(tempx, tempy);
        cGamepad.downButtons.down && !cPlayer._crashed && ctx.fill();
        ctx.stroke();

        //change this to change the way the symbol is drawn
        if (cGamepad.downButtons.down && !cPlayer._crashed) {
            ctx.strokeStyle = ldcolor;
        }
        ctx.beginPath();
        ctx.lineTo(tempx + (1 * bsize) / 6, tempy + (1 * bsize) / 3);
        ctx.lineTo(tempx + (1 * bsize) / 2, tempy + (2 * bsize) / 3);
        ctx.lineTo(tempx + (5 * bsize) / 6, tempy + (1 * bsize) / 3);
        ctx.stroke();

        //right key
        ctx.strokeStyle = lcolor;
        tempx = sx + 2 * bsize;
        tempy = sy + bsize;
        ctx.beginPath();
        ctx.lineTo(tempx, tempy);
        ctx.lineTo(tempx, tempy + bsize);
        ctx.lineTo(tempx + bsize, tempy + bsize);
        ctx.lineTo(tempx + bsize, tempy);
        ctx.lineTo(tempx, tempy);
        cGamepad.downButtons.right && !cPlayer._crashed && ctx.fill();
        ctx.stroke();

        //change this to change the way the symbol is drawn
        if (cGamepad.downButtons.right && !cPlayer._crashed) {
            ctx.strokeStyle = ldcolor;
        }
        ctx.beginPath();
        ctx.lineTo(tempx + (1 * bsize) / 3, tempy + (1 * bsize) / 6);
        ctx.lineTo(tempx + (2 * bsize) / 3, tempy + (1 * bsize) / 2);
        ctx.lineTo(tempx + (1 * bsize) / 3, tempy + (5 * bsize) / 6);
        ctx.stroke();
    }
}
var track = 0;
window.setInterval(function () {
    if ($('#track-data').data('t_id') != track) {
        track = $('#track-data').data('t_id');
        function rInterval() {
            window.clearInterval(v);
        }
        var v = window.setInterval(function () {
            if (GameManager != undefined && GameManager.game != undefined) {
                rInterval();
                createjs.Ticker.addEventListener('tick', drawDisplay);
            }
        }, 250);
    }
}, 500);


Star Finder Script.user.js
// ==UserScript==
// @name         Star Finder Script
// @namespace    http://tampermonkey.net/
// @version      0.2
// @description  creates arrows pointing to the stars in a track outside of where they can be seen
// @author       Pie42
// @match        https://www.freeriderhd.com/*
// @grant        none
// ==/UserScript==

var track = 0,
    eListener,
    //the base color of the arrows
    //feel free to change it
    color = [0, 255, 0];
function mod(m, n) {
    return ((m % n) + n) % n;
}
function main() {
    var canvas = GameManager.game.canvas,
        ctx = canvas.getContext('2d');
    eListener = createjs.Ticker.addEventListener('tick', function() {
        var camera = GameManager.game.currentScene.camera,
            fIndex = camera.focusIndex,
            cPlayer = GameManager.game.currentScene.playerManager._players[fIndex],
            tC = cPlayer._powerupsConsumed.targets,
            cPos = camera.position,
            cZoom = camera.zoom * 2,
            cBounds = [cPos.x - (canvas.width / cZoom), cPos.y - (canvas.height / cZoom), cPos.x + (canvas.width / cZoom), cPos.y + (canvas.height / cZoom)],
            rect = Math.atan2(canvas.width, canvas.height),
            targets = GameManager.game.currentScene.track.targets,
            done = 0;
        for (let i in targets) {
            let j = targets[i];
            //this 500 is only here because for my device, 500 is about the highest number of arrows it can render before lag starts to happen.
            //however, i am using a chromebook with 2 gb of ram, so if your device can do better, feel free to change this number.
            //         vvv
            if (done > 500) break;
            if (tC.indexOf(j.id) == -1) {
                done++;
                if (j.x < cBounds[0] || j.y < cBounds[1] || j.x > cBounds[2] || j.y > cBounds[3]) {
                    let angle = Math.atan2(j.y - cPos.y, j.x - cPos.x),
                        sDist = Math.sqrt((j.y - cPos.y) ** 2 + (j.x - cPos.x) ** 2) / cZoom,
                        size = (1000 / cZoom) / Math.sqrt(sDist),
                        nAngle = mod(angle, Math.PI),
                        nHeight = canvas.height / 2,
                        nWidth = canvas.width / 2,
                        sine = Math.sin(angle),
                        cosine = Math.cos(angle),
                        gHeight = nHeight - Math.abs(size * sine),
                        gWidth = nWidth - Math.abs(size * cosine),
                        dy = (sine > 0) ? gHeight : -1 * gHeight,
                        dx = (cosine > 0) ? gWidth : -1 * gWidth;
                    if (Math.abs(dx * sine) < Math.abs(dy * cosine)) {
                        dy = (dx * sine) / cosine;
                    }
                    else {
                        dx = (dy * cosine) / sine;
                    }
                    let pos = {x: dx + nWidth, y: dy + nHeight},
                        //i really hate the 2.1 here because it's just sort of a 'magic number'
                        //that makes the opacity work pretty well in both normal view and full screen.
                        //if there are numbers that work better for you, feel free to change it.
                        //alternatively, if you want just one opacity, you can replace it with:
                        //alpha = 0.5;
                        //           vvv
                        //alpha = 1 - ((2.1 / cZoom) / (sDist / (Math.sqrt(dy ** 2 + dx ** 2) / cZoom)));
                        alpha = 1 - ((2.1 / cZoom) / (sDist / (Math.sqrt(dy ** 2 + dx ** 2) / cZoom)));
                    ctx.fillStyle = `rgba(${color.join(', ')}, ${alpha.toString()})`
                    //this is where the arrows themselves are drawn
                    //if you want to draw something else besides arrows, go ahead!
                    //pos.x and pos.y are the center of the arrows
                    ctx.beginPath();
                    ctx.lineTo(pos.x + (Math.cos(angle) * size), pos.y + (Math.sin(angle) * size));
                    ctx.lineTo(pos.x + (Math.cos(angle + (2 * Math.PI / 3)) * size), pos.y + (Math.sin(angle + (2 * Math.PI / 3)) * size));
                    ctx.lineTo(pos.x + (Math.cos(angle - (2 * Math.PI / 3)) * size), pos.y + (Math.sin(angle - (2 * Math.PI / 3)) * size));
                    ctx.fill();
                }
                else {

                }
            }
        }
    })
}
window.setInterval(function () {
    if ($("#track-data").data("t_id") != track) {
        track = $("#track-data").data("t_id");
        function rInterval() {
            window.clearInterval(v)
        }
        var v = window.setInterval(function() {
            if (GameManager != undefined && GameManager.game != undefined) {
                rInterval();
                main();
            }
        }, 250)
    }
}, 500)

FRHD Input Display 2.user.js
// ==UserScript==
// @name         FRHD Input Display 2
// @namespace    http://freeriderhd.com/
// @version      .9
// @description  try to take over the world!
// @author       Pie42
// @include      *freeriderhd.com*
// @grant        none
// ==/UserScript==

/*
  Short explanation: The way this works is that the main drawing function first hooks itself up to the game's main update function,
  because that's the best way to make it draw correctly. After that, every time the game ticks, the input display is drawn onto the game canvas.
  To get button highlighting, each drawing function checks the current player's '_gamepad.downButtons' to see if the current button is being held down
  (whether it's by the current player or the current replay). If the current button is being held down, the box is filled in.

  This input display will automatically highlight the current player, meaning that if you're watching a ghost, the display will show the ghost's
  inputs, and if you're playing, it will show your inputs. I decided not to include 'enter' and 'backspace' because it's pretty easy to tell if those
  are being held. Enjoy!

  Note that you can customize this to your heart's content. I have indicated the main customization things, but feel free to change whatever(just don't
  blame me if it breaks).
*/
//var difx = 0,
//    dify = 0;
function spaceTest(e) {
    if (e.key == ' ') space = e.type == 'keydown' ? true : false;
}
window.addEventListener('keydown', spaceTest);
window.addEventListener('keyup', spaceTest);
var spin = 0,
    up = 0,
    down = 0,
    left = 0,
    right = 0,
    sAcc = 0,
    space = false;
function circle(ctx, x, y, r) {
    ctx.beginPath();
    for (let i = 0; i <= Math.PI * 2; i += Math.PI / 20) {
        ctx.lineTo(x + (Math.cos(i) * r), y + (Math.sin(i) * r));
    }
    //ctx.stroke();
}
function symbol(ctx, tempx, tempy, bsize, diff) {
    ctx.beginPath();
    ctx.lineTo(tempx + (Math.cos(spin + (diff)) * (bsize / 6)), tempy + (Math.sin(spin + (diff)) * (bsize / 6)));
    ctx.lineTo(tempx + (Math.cos(spin + ((2 * Math.PI / 3) + diff)) * (bsize / 6)), tempy + (Math.sin(spin + ((2 * Math.PI / 3) + diff)) * (bsize / 6)));
    ctx.lineTo(tempx + (Math.cos(spin + ((4 * Math.PI / 3) + diff)) * (bsize / 6)), tempy + (Math.sin(spin + ((4 * Math.PI / 3) + diff)) * (bsize / 6)));
    ctx.lineTo(tempx + (Math.cos(spin + (diff)) * (bsize / 6)), tempy + (Math.sin(spin + (diff)) * (bsize / 6)));
    ctx.stroke();
}
function drawDisplay() {
    var container = document.getElementById('game-container'),
        children = (container == void 0) ? undefined : container.children,
        canvas = (children == void 0) ? undefined : children[1];
    if (canvas != void 0){
        var ctx = canvas.getContext('2d');
        var cPlayer = GameManager.game.currentScene.playerManager._players[GameManager.game.currentScene.camera.focusIndex],
            cGamepad = cPlayer._gamepad,
            h = canvas.height,
            w = canvas.width;

        //main customization variables
        //to customize a symbol, find the comment stating that key(keys are listed left to right and top to bottom), and then change the drawing done there.
        var bsize = h / 10,     //size of the boxes
            dist = h / 8,      //distance from the edge of the screen
            lcolor = '#000000', //color of the lines
            fcolor = '#1B5264', //color that currently pressed inputs are filled in
            lwidth = 2.5;       //width of the lines

        //sy and sx can also be customized. right now, they are set to put the input display in the lower left corner, but to move it to the top, you can use:
        //var sy = dist,
        //instead of the line that is currently there.
        var sy = (h - dist) - (bsize),

            //to move the display to the right, you can use
            //sx = (w - dist) - (3 * bsize),
            //instead of the line that is currently there.
            sx = dist + bsize,
            tempx, tempy;
        ctx.strokeStyle = lcolor;
        ctx.fillStyle = fcolor;
        ctx.lineWidth = lwidth;

        //GameManager.game.currentScene.restartTrack && (up = 0, down = 0, left = 0, right = 0, spin = Math.PI, sAcc = 0);


        //z key
        /*tempx = sx + (2 * bsize);
        tempy = sy;
        ctx.beginPath();
        ctx.lineTo(tempx, tempy);
        ctx.lineTo(tempx, tempy + bsize);
        ctx.lineTo(tempx + bsize, tempy + bsize);
        ctx.lineTo(tempx + bsize, tempy);
        ctx.lineTo(tempx, tempy);*/
        tempx = sx;
        tempy = sy;
        circle(ctx, tempx, tempy, bsize / 2);
        cGamepad.downButtons.z && !cPlayer._crashed && (ctx.fill(), sAcc += Math.PI / 100);
        ctx.stroke();
        tempx -= bsize / 2;
        tempy -= bsize / 2;
        //change this to change the way the symbol is drawn
        ctx.beginPath();
        ctx.lineTo(tempx + (bsize / 3), tempy + (bsize / 3));
        ctx.lineTo(tempx + (2 * bsize / 3), tempy + (bsize / 3));
        ctx.lineTo(tempx + (bsize / 3), tempy + (2 * bsize / 3));
        ctx.lineTo(tempx + (2 * bsize / 3), tempy + (2 * bsize / 3));
        ctx.stroke();


        //up arrow key
        tempx = sx + (Math.cos(spin + (Math.PI / 2)) * (bsize + up));
        tempy = sy + (Math.sin(spin + (Math.PI / 2)) * (bsize + up));
        circle(ctx, tempx, tempy, bsize / 2);
        cGamepad.downButtons.up && !cPlayer._crashed && (ctx.fill(), up++);
        //cGamepad.downButtons.pause && !cPlayer._crashed && (ctx.fill(), up++);
        //space && !cPlayer._crashed && (ctx.fill(), up++);
        ctx.stroke();

        //change this to change the way the symbol is drawn
        symbol(ctx, tempx, tempy, bsize, Math.PI / 2);


        //left key
        tempx = sx + (Math.cos(spin + (0)) * (bsize + left));
        tempy = sy + (Math.sin(spin + (0)) * (bsize + left));
        circle(ctx, tempx, tempy, bsize / 2);
        cGamepad.downButtons.left && !cPlayer._crashed && (ctx.fill(), left++);
        ctx.stroke();

        //change this to change the way the symbol is drawn
        symbol(ctx, tempx, tempy, bsize, 0);


        //down key
        tempx = sx + (Math.cos(spin + (3 * Math.PI / 2)) * (bsize + down));
        tempy = sy + (Math.sin(spin + (3 * Math.PI / 2)) * (bsize + down));
        circle(ctx, tempx, tempy, bsize / 2);
        cGamepad.downButtons.down && !cPlayer._crashed && (ctx.fill(), down++);
        ctx.stroke();

        //change this to change the way the symbol is drawn
        symbol(ctx, tempx, tempy, bsize, 3 * Math.PI / 2);


        //right key
        tempx = sx + (Math.cos(spin + (Math.PI)) * (bsize + right));
        tempy = sy + (Math.sin(spin + (Math.PI)) * (bsize + right));
        circle(ctx, tempx, tempy, bsize / 2);
        cGamepad.downButtons.right && !cPlayer._crashed && (ctx.fill(), right++);
        ctx.stroke();

        //change this to change the way the symbol is drawn
        symbol(ctx, tempx, tempy, bsize, Math.PI);
        /*difx = Math.max(Math.min(difx, 20), -20);
        difx *= 0.95;
        dify = Math.max(Math.min(dify, 20), -20);
        dify *= 0.95;*/
        spin += sAcc;
        sAcc *= 0.005;
        up *= .95;
        down *= .95;
        left *= .95;
        right *= .95;
    }
}
var track = 0;
window.setInterval(function () {
    if ($("#track-data").data("t_id") != track) {
        track = $("#track-data").data("t_id");
        function rInterval() {
            window.clearInterval(v)
        }
        var v = window.setInterval(function() {
            if (GameManager != undefined && GameManager.game != undefined && GameManager.game.currentScene != undefined) {
                rInterval();
                createjs.Ticker.addEventListener("tick", drawDisplay);
                spin = Math.PI;
                GameManager.game.currentScene.restart = function() {
                    GameManager.game.currentScene.settings.mobile ? GameManager.game.currentScene.message.show("Press Any Button To Start", 1, "#333333") : GameManager.game.currentScene.message.show("Press Any Key To Start", 1, "#333333", "#FFFFFF"),
                    GameManager.game.currentScene.track.resetPowerups(),
                    GameManager.game.currentScene.restartTrack = !1,
                    GameManager.game.currentScene.state.paused = !1,
                    GameManager.game.currentScene.state.playing = !GameManager.game.currentScene.settings.waitForKeyPress,
                    GameManager.game.currentScene.ticks = 0,
                    GameManager.game.currentScene.playerManager.reset(),
                    GameManager.game.currentScene.playerManager.getPlayerCount() > 0 && (GameManager.game.currentScene.camera.focusIndex = 1),
                    GameManager.game.currentScene.camera.focusOnPlayer(),
                    GameManager.game.currentScene.camera.fastforward(),
                    GameManager.game.currentScene.showControlPlanel("main");
                    up = 0, down = 0, left = 0, right = 0, spin = Math.PI, sAcc = 0;
                }
            }
        }, 250)
    }
}, 500)